File: Binary Search/BinarySearch.cpp
#include "BinarySearch.hpp"

// binary search

int binarySearch(std::vector<int> array, int target) {
  
  int midIndex = (int) array.size() / 2;
  int mid = array[midIndex];
  if (mid == target) return midIndex;
  if (array.size() == 1 && mid != target) return -1;
  
  std::vector<int> left (array.begin(), array.end() - midIndex);
  std::vector<int> rite (array.begin() + midIndex + 1, array.end());
  
  if (mid > target) {
    return binarySearch(left, target);
  } else {
    int index = binarySearch(rite, target);
    return index < 0 ? index : midIndex + index + 1;
  }
}

// debugger

void print(std::vector<int> array) {
  std::cout << "{ ";
  for (int index = 0; index < array.size(); index++) {
    std::cout << array[index];
    if (index < array.size() - 1) std::cout << ", ";
  }
  std::cout << " }\n" << std::endl;
}

File: Binary Search/binary_search.rb
  def bsearch (target_value, &prc)
      prc ||= Proc.new { |a, b| a <=> b }

      mid_index = self.length / 2
      mid_value = self[mid_index]
      return mid_index if mid_value == target_value

      return nil if self.length <= 1
      left = self.take(mid_index)
      rite = self.drop(mid_index)

      case prc.call(target_value, mid_value)
      when  1
        index = rite.bsearch(target_value, &prc)
        return mid_index + index if index
      when -1
        return left.bsearch(target_value, &prc)
      end
  end